File: Review-09-Java-Shapes/Example-6/src/test/java/edu/odu/cs/cs330/examples/shapes/TestSquare.java
package edu.odu.cs.cs330.examples.shapes;

import org.junit.jupiter.api.TestMethodOrder;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.BeforeEach;

import static org.junit.jupiter.api.Assertions.*;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import org.hamcrest.core.IsNull;
import static org.hamcrest.number.IsCloseTo.closeTo;

import java.io.StringReader;
import java.util.Scanner;

/**
 * 1 - Does this piece of code perform the operations
 *     it was designed to perform?
 *
 * 2 - Does this piece of code do something it was not
 *     designed to perform?
 *
 * 1 Test per mutator
 */
@TestMethodOrder(MethodOrderer.MethodName.class)
public class TestSquare {

    private Square generic;
    private Square fancy;

    @BeforeEach
    public void setUp()
    {
        this.generic = new Square();
        this.fancy   = new Square(2);
    }

    @Test
    public void testDefaultConstructor()
    {
        assertThat(this.generic.name(), equalTo("Square"));
        assertThat(this.generic.side(), closeTo(1, 1e-8));
    }

    @Test
    public void testConstructor()
    {
        assertThat(this.fancy.name(), equalTo("Square"));
        assertThat(this.fancy.side(), closeTo(2.0, 1e-8));
    }

    @Test
    public void testSideSetter()
    {
        Square aSquare = (Square) new Square();

        aSquare.side(7.39);

        assertThat(aSquare.side(), closeTo(7.39, 1e-8));
        assertThat(aSquare.name(), equalTo("Square"));
    }

    @Test
    public void testArea()
    {
        assertThat(this.generic.area(),
                   closeTo(Math.pow(this.generic.side(), 2), 1e-8));

        assertThat(this.fancy.area(),
                   closeTo(Math.pow(this.fancy.side(), 2), 1e-8));
    }

    @Test
    public void testPerimeter()
    {
        assertThat(this.generic.perimeter(),
                   closeTo(4 * this.generic.side(), 1e-8));

        assertThat(this.fancy.perimeter(),
                   closeTo(4 * this.fancy.side(), 1e-8));
    }

    @Test
    public void testClone()
        throws CloneNotSupportedException
    {
        Square aCopy = (Square) this.fancy.clone();

        assertThat(aCopy, is(not(sameInstance(this.fancy))));

        // I really should have defined __eq__
        assertThat(aCopy.side(), closeTo(this.fancy.side(), 0.001));
    }

    @Test
    public void testToString()
    {
        String fancyStr = this.fancy.toString();

        assertThat(fancyStr, startsWith("Name"));
        assertThat(fancyStr, containsString("Square"));
        assertThat(fancyStr,
                   containsString(String.format("%-12s: %24.4f",
                                                "Perimeter",
                                                this.fancy.perimeter())));

        assertThat(fancyStr,
                    containsString(String.format("%-12s: %24.4f",
                                                 "Area",
                                                  this.fancy.area())));

        assertThat(fancyStr,
                    containsString(String.format("%-12s: %24.4f",
                                                 "Side",
                                                 this.fancy.side())));

        assertThat(fancyStr, endsWith("\n"));
    }

    @Test
    public void testRead()
    {
        Scanner snr = new Scanner(new StringReader("9.7"));

        Square aSquare = new Square();
        aSquare.read(snr);

        assertThat(aSquare.name(), equalTo("Square"));
        assertThat(aSquare.side(), closeTo(9.7, 0.01));
    }
}

File: Review-06-Python-Shapes/Example-7/tests/shapes/test_square.py
from hamcrest import *
import unittest

from shapes.square import Square
from shapes.shape import (Shape, FPT_FMT)

import copy


class TestSquare(unittest.TestCase):
    """
    1 - Does this piece of code perform the operations
        it was designed to perform?

    2 - Does this piece of code do something it was not
        designed to perform?

    1 Test per mutator
    """

    def setUp(self):
        self.generic = Square()
        self.fancy = Square(side=2)

    def test_default_constructor(self):
        assert_that(self.generic.name, equal_to("Square"))
        assert_that(self.generic.side, close_to(1, 1e-8))

    def test_constructor(self):
        assert_that(self.fancy.name, equal_to("Square"))
        assert_that(self.fancy.side, close_to(2.0, 1e-8))

    def test_side_setter(self):
        a_square = Square()

        a_square.side = 7.39

        assert_that(a_square.side, close_to(7.39, 1e-8))
        assert_that(a_square.name, equal_to("Square"))

    def test_area(self):
        assert_that(self.generic.area(),
                    close_to(self.generic.side ** 2, 1e-8))

        assert_that(self.fancy.area(),
                    close_to(self.fancy.side ** 2, 1e-8))

    def test_perimeter(self):
        assert_that(self.generic.perimeter(),
                    close_to(4 * self.generic.side, 1e-8))

        assert_that(self.fancy.perimeter(),
                    close_to(4 * self.fancy.side, 1e-8))

    def test_deep_copy(self):
        a_copy = copy.deepcopy(self.fancy)

        assert_that(a_copy, is_not(same_instance(self.fancy)))

        # I really should have defined __eq__
        assert_that(a_copy.side, close_to(self.fancy.side, 0.001))

    def test_str(self):
        fancy_str = str(self.fancy)

        assert_that(fancy_str, starts_with("Name"))
        assert_that(fancy_str, contains_string("Square"))
        assert_that(fancy_str,
                    contains_string(FPT_FMT.format("Perimeter",
                                                   self.fancy.perimeter())))
        assert_that(fancy_str,
                    contains_string(FPT_FMT.format("Area",
                                                   self.fancy.area())))
        assert_that(fancy_str,
                    contains_string(FPT_FMT.format("Side",
                                                   self.fancy.side)))
        assert_that(fancy_str, ends_with("\n"))
